File: python/Brunel/_business.py

__all__ = ["Business"]


def _setState(state, val, default=None):
    try:
        result = state[val]
        if result:
            return result
        else:
            return default
    except KeyError:
        return default


def _mergeProjects(old, new, key):
    old = old[key]
    new = new[key]

    for project, values in new.items():
        if project in old:
            for value in values:
                if value not in old[project]:
                    old[project].append(value)
        else:
            old[project] = values


def _mergeStateItems(old, new, key):
    """ Merge two state items, adds items from new to old state

        Args:
            old (dict): Old object's state
            new (dict): New object's state
            key (str): Key to acccess dictionary values
        Returns:
            None
    """
    for id, data in new[key].items():
        if key not in old:
            old[key] = {}

        old[key][id] = data


class Business:
    """Holds information about a Business or Institution in the network"""

    def __init__(self, props=None, getHook=None):
        self._getHook = getHook

        self.state = {
            "name": None,
            "id": None,
            "projects": {},
            "sources": [],
            "scores": {},
            "positions": {},
            "affiliations": {},
            "highlighted": {},
            "notes": [],
            "weight": {},
            "weight_path": {},
        }

        self.setState(props)

    def __str__(self):
        return f"Business({self.getName()})"

    def getID(self):
        return self.state["id"]

    def getName(self):
        return self.state["name"]

    def getAffiliations(self):
        result = []

        affiliations = self.state["affiliations"]

        for affiliation in affiliations.keys():
            result.append((self._getHook(affiliation), affiliations[affiliation]))

        return result

    def getPositions(self):
        result = {}

        positions = self.state["positions"]

        for project in positions.keys():
            if project not in result:
                result[project] = []

            for position in positions[project]:
                result[project].append(self._getHook(position))

        return result

    def setState(self, state):
        if not state:
            return

        self.state["name"] = _setState(state, "name")
        self.state["id"] = _setState(state, "id")
        self.state["projects"] = _setState(state, "projects", {})
        self.state["affiliations"] = _setState(state, "affiliations", {})
        self.state["scores"] = _setState(state, "scores", {})
        self.state["sources"] = _setState(state, "sources", [])
        self.state["highlighted"] = _setState(state, "highlighted", {})
        self.state["notes"] = _setState(state, "notes", [])
        self.state["weight"] = _setState(state, "weight")
        self.state["weight_path"] = _setState(state, "weight_path", {})
        self.state["positions"] = _setState(state, "positions", {})

        if self.state["name"] == "None" or self.state["name"] is None:
            raise ValueError(f"No name for {self}?")

    def merge(self, other):
        """ Merge two businesses together. This function is used when merging in matching
            people that are involved in another project.

            Args:
                other (Business): Business to merge with this Business object
            Returns:
                Business: New Business object created from combined states
        """
        import copy as _copy

        state = _copy.copy(self.state)

        _mergeProjects(state, other.state, "positions")
        _mergeProjects(state, other.state, "affiliations")
        _mergeProjects(state, other.state, "sources")
        _mergeProjects(state, other.state, "highlighted")

        # for project, dates in other.state["projects"].items():
        #     state["projects"][project] = dates

        # for id, weight in other.state["weight"].items():
        #     state["weight"][id] = weight

        _mergeStateItems(state, other.state, "projects")
        _mergeStateItems(state, other.state, "weight")
        _mergeStateItems(state, other.state, "weight_path")

        # for id, dates in other.state["projects"].items():
        #     state["projects"][id] = dates

        # for id, weight in other.state["weight"].items():
        #     state["weight"][id] = weight

        b = Business()
        b.state = state
        b._getHook = self._getHook

        return b

    def toDry(self):
        return self.state

    @staticmethod
    def unDry(value):
        return Business(value)

    @staticmethod
    def load(data):
        return Business(data)

File: src/model/Business.js
import Dry from "json-dry";
import lodash from "lodash";

import WeightPath from "./WeightPath";
import DateRange from "./DateRange";

import { ValueError } from "./Errors";

function setState(val, def = null) {
  if (val) {
    return val;
  } else {
    return def;
  }
}

function _filterWindow(values, window) {
  if (!values) {
    return values;
  }

  let ret = null;

  Object.keys(values).forEach((key) => {
    let dates = values[key];

    let intersect = window.intersect(dates);

    if (!intersect) {
      if (!ret) {
        ret = { ...values };
      }
      delete ret[key];
    }
  });

  if (ret) {
    return ret;
  } else {
    return values;
  }
}

function _filterProject(values, project) {
  if (!values) {
    return values;
  }

  let ret = null;

  Object.keys(values).forEach((key) => {
    if (!(key in project)) {
      if (!ret) {
        ret = { ...values };
      }
      delete ret[key];
    }
  });

  if (ret) {
    return ret;
  } else {
    return values;
  }
}

class Business {
  constructor(props) {
    this.state = {
      name: null,
      id: null,
      projects: {},
      sources: {},
      positions: {},
      scores: {},
      affiliations: {},
      notes: [],
      weight: {},
      weight_path: {},
      is_highlighted: false,
      is_selected: false,
    };

    this.setState(props);

    this._getHook = null;
    this._isABusinessObject = true;
  }

  static clone(item) {
    let c = new Business();
    c._getHook = item._getHook;
    c.state = lodash.cloneDeep(item.state);
    return c;
  }

  isNode() {
    return true;
  }

  inGroup(group) {
    if (group.getID) {
      let ids = {};
      ids[group.getID()] = 1;
      group = ids;
    }

    let seen = {};

    Object.keys(this.state.affiliations).forEach((key) => {
      for (let index in this.state.affiliations[key]) {
        if (this.state.affiliations[key][index] in group) {
          seen[this.state.affiliations[key][index]] = 1;
        }
      }
    });

    Object.keys(this.state.positions).forEach((key) => {
      for (let index in this.state.positions[key]) {
        if (this.state.positions[key][index] in group) {
          seen[this.state.positions[key][index]] = 1;
        }
      }
    });

    return Object.keys(seen).length === Object.keys(group).length;
  }

  getID() {
    return this.state.id;
  }

  filterKeyDate(keydate){
    const weightpath = this.getWeightPath();

    let weight = weightpath.getWeightAtDate(keydate);

    if (weight <= 0.0){
      return null;
    }

    this.state.weight = weight;

    return this;
  }

  filterSource(source) {
    if (source.getID) {
      let id = source.getID();
      source = {};
      source[id] = 1;
    }

    let nsources = Object.keys(source).length;

    let seen = {};

    Object.keys(this.state.sources).forEach((key) => {
      let s = this.state.sources[key];

      Object.keys(source).forEach((source_id) => {
        if (s.includes(source_id)) {
          seen[source_id] = 1;
        }
      });
    });

    if (Object.keys(seen).length !== nsources) {
      return null;
    } else {
      return this;
    }
  }

  filterProject(project) {
    if (project.getID) {
      let id = project.getID();
      project = {};
      project[id] = 1;
    }

    let nprojects = Object.keys(project).length;

    let seen = {};
    let new_projects = {};

    Object.keys(this.state.projects).forEach((key) => {
      if (key in project) {
        seen[key] = 1;
        new_projects[key] = this.state.projects[key];
      }
    });

    if (Object.keys(seen).length !== nprojects) {
      return null;
    }

    let affiliations = _filterProject(this.state.affiliations, project);
    let positions = _filterProject(this.state.positions, project);
    let weight = _filterProject(this.state.weight, project);
    let sources = _filterProject(this.state.sources, project);

    if (affiliations !== this.state.affiliations || positions !== this.state.positions ||
        weight !== this.state.weight || sources !== this.state.sources) {
      let business = new Business();
      business.state = { ...this.state };
      business.state.affiliations = affiliations;
      business.state.positions = positions;
      business.state.sources = sources;
      business.state.weight = weight;
      business.state.projects = new_projects;
      business._getHook = this._getHook;
      return business;
    } else {
      return this;
    }
  }

  filterWindow(window) {
    if (!window) {
      return this;
    } else if (!window._isADateRangeObject) {
      window = new DateRange(window);
    }

    let affiliations = _filterWindow(this.state.affiliations, window);
    let positions = _filterProject(this.state.positions, window);

    if (affiliations !== this.state.affiliations || positions !== this.state.positions) {
      let business = new Business();
      business.state = { ...this.state };
      business.state.affiliations = affiliations;
      business.state.positions = positions;
      business._getHook = this._getHook;
      return business;
    } else {
      return this;
    }
  }

  setState(state) {
    if (state) {
      this.state.name = setState(state.name);
      this.state.id = setState(state.id);
      this.state.projects = setState(state.projects, {});
      this.state.affiliations = setState(state.affiliations, {});
      this.state.positions = setState(state.positions, {});
      this.state.scores = setState(state.scores, {});
      this.state.sources = setState(state.sources, {});
      this.state.notes = setState(state.notes, []);
      this.state.weight = setState(state.weight);
      this.state.weight_path = setState(state.weight_path, {});

      if (!this.state.name) {
        throw new ValueError("You cannot have an Business without a name");
      }
    }
  }

  _updateHooks(hook) {
    this._getHook = hook;
  }

  merge() {
    return this;
  }

  getProjectID() {
    // return the first matching project ID
    return Object.keys(this.state.projects)[0];
  }

  getWeight(project_id = null) {
    let weight = this.state.weight;

    if (weight === null){
      weight = 5.0;
    }

    return weight;
  }

  getWeightPath(project_id = null) {
    if (project_id === null) {
      // use the first project's weight
      project_id = Object.keys(this.state.weight_path)[0];
    }

    const weight_path = this.state.weight_path[project_id];

    if (!weight_path){
      return new WeightPath();
    }

    return weight_path;
  }

  toString() {
    return `Business(${this.getName()})`;
  }

  getName() {
    return this.state.name;
  }

  getInitials() {
    var parts = [];

    this.state.name.split(" ").forEach((n) => {
      parts.push(n[0]);
    });

    if (parts.length > 0) {
      return parts.join("");
    } else {
      return "??";
    }
  }

  getAffiliations() {
    return this.state.affiliations;
  }

  getPosition(projectID) {
    // Return the position for the associated projectID
    // Returns an array of position IDs
    return this.state.positions[projectID];
  }

  getPositions() {
    // Return the position for the associated projectID
    // Returns an array of position IDs
    return this.state.positions;
  }

  getSources() {
    return this.state.sources;
  }

  getScores() {
    return this.state.scores;
  }

  setSelected(val) {
    if (val) {
      this.state.is_selected = true;
    }
    else {
      this.state.is_selected = false;
    }
  }

  getSelected() {
    return this.state.is_selected;
  }

  setHighlighted(val) {
    if (val) {
      this.state.is_highlighted = true;
    }
    else {
      this.state.is_highlighted = false;
    }
  }

  getHighlighted() {
    return this.state.is_highlighted;
  }

  isNonContributingEngineer(project_id = null) {
    if (project_id === null) {
      // return this status with the first project
      project_id = Object.keys(this.state.positions)[0];
    }

    const positions = this.state.positions[project_id];

    for (let i = 0; i < positions.length; ++i){
      const position = this._getHook(positions[i]);
      const name = position.getCanonical();
      if (name.includes("non-cont")) {
        return true;
      }
    }

    return false;
  }

  getNode() {
    let node = {
      id: this.getID(),
      label: this.getName(),
      title: this.getName(),
      initials: this.getInitials(),
      shape: "square",
      highlighted: this.getHighlighted(),
      selected: this.getSelected(),
      weight: this.getWeight(),
      type: "business",
    };

    return node;
  }

  toDry() {
    return { value: this.state };
  }
}

Business.unDry = function (value) {
  return new Business(value);
};

Dry.registerClass("Business", Business);

export default Business;
